src/azul/service/responseobjects/storage_service.py

```diff
@@ -37,14 +43,144 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
 
         return object_key
 
+    @contextmanager
+    def multipart_upload(self, object_key: str, content_type: str):
+        api_response = self.client.create_multipart_upload(Bucket=self.__bucket_name,
+                                                           Key=object_key,
+                                                           ContentType=content_type)
+        upload_id = api_response['UploadId']
+
+        with ThreadPoolExecutor(max_workers=MULTIPART_UPLOAD_MAX_WORKERS) as thread_pool:
+            handler = MultipartUploadHandler(self.__bucket_name, object_key, upload_id, thread_pool)
+
+            try:
+                yield handler
+            except Exception as e:
+                logger.error('Upload %s: Aborting due to unexpected error', upload_id)
+                handler.abort()
+                logger.warning('Upload %s: Aborted', upload_id)
+                raise UnexpectedMultipartUploadAbort(f'{type(e).__name__}: {e}')
+
+            if handler.is_active:
+                handler.complete()
+
     def get_presigned_url(self, key: str) -> str:
         return self.client.generate_presigned_url(ClientMethod='get_object',
                                                   Params=dict(Bucket=self.__bucket_name, Key=key))
 
     def create_bucket(self, bucket_name: str = None):
         self.client.create_bucket(Bucket=(bucket_name or self.__bucket_name))
-        logger.warning(f'{type(self).__name__}: Created a bucket called "{bucket_name or self.__bucket_name}"')
+
+
+class MultipartUploadHandler:
+
+    def __init__(self, bucket_name, object_key, upload_id, thread_pool):
+        self.__resource = boto3.resource('s3')
+        self.__bucket_name = bucket_name
+        self.__object_key = object_key
+        self.__upload_id = upload_id
+        self.__handler = self.__resource.MultipartUpload(self.__bucket_name, self.__object_key, self.__upload_id)
+        self.__next_part_number = 1
+        self.__closed = False
+        self.__parts = []
+        self.__futures = []
+        self.__thread_pool = thread_pool
+        self.__new_part_lock = Lock()
+        self.__sync_lock = Lock()
```

Comment:
It's usually a bad smell to use locks conjunction with ThreadPoolExecutor. The whole point of the `concurrent` package is provide an abstraction that lets you write concurrent code WITHOUT having to manage the locks yourself.

These locks can be avoided if you make _upload_part return the part and then use future.get() to get the return value.

Reply:
So, if I understand you correctly, in the `push` method, instead of using locks, it will be something like:

```python
def push(self, data: bytes):
    part = self.thread_pool.submit(self.resume_last_part).result()
    if not part.is_uploadable:
        return
    part.uploaded = True
    self.__futures.append(self.__thread_pool.submit(self._upload_part, part))
```

where that part `self.thread_pool.submit(self.resume_last_part).result()` is safe enough to guarantee that the we will either resume the last part, which is previously not uploadable, or create a new part without race condition, right?

Reply:
I don't know what you mean by "resuming" a part. Once you have enough data for a part you should offload uploading that part as a task in the thread pool. Once all parts have been uploaded (as indicated by all futures indicating success), the MP upload can be completed by the main thread.